clusterversion, build: mint 23.2 & init 24.1 development

celiala opened the pull request:
As part of the 23.2 stability period tasks, this PR preps the 23.2 release branch and initializes 24.1 development, as per this release runbook: [Prep Release and Define Start of Development for Next Release](https://cockroachlabs.atlassian.net/wiki/spaces/RE/pages/2924740768/Prep+Release+and+Define+Start+of+Development+for+Next+Release).

### Important Notes:

- The commits in this PR _**must**_ be done as a single, atomic PR to `master`, so that `master` always has `developmentBranch` set to `false`:

   - In these sequence of commits, we temporarily set `developmentBranch` to `false`, which is backported to `release-23.2`. We then set `developmentBranch` back to `true` after adding `V24_1Start`.

- The below timing for merging backports must be followed (additional context below):

|commit|backport to release-23.2?|when to merge backport|
|-|-|-|
|1. clusterversion: set developmentBranch to false|yes|**before** selecting beta.1 candidate|
|2. build: advance version to beta.1|yes|**before** selecting beta.1 candidate|
|3. clusterversion: mint 23.2|yes|should be the **final backport before the final RC**|
|4. clusterversion: define start of 24.1 development and placeholder key|no|_(not backported)_|


### Context around the particular timing of these backports

This section gives additional context around why we need to backport certain commits at various times.

#### Commits to be backported **before** selecting beta.1 candidate:

- clusterversion: set developmentBranch to false
- build: advance version to beta.1

Why: In order for betas to be upgradeable to future betas and beyond, developmentBranch must be set to false

#### Minting should not be backported until **just before before the final RC**
Commits:
- clusterversion: mint 23.2

Why: minting the release branch sets the release as 23.2. Minting also prevents the addition of subsequent gates, though, so we shouldn't merge this until we're _sure_ that we no longer need additional 23.2 gates.


Fixes: https://github.com/cockroachdb/cockroach/issues/111889

Release note: None.
Epic: REL-437.
(https://cockroachlabs.atlassian.net/browse/REL-437)

celiala committed:
clusterversion: mint 23.2

This mints the release branch as 23.2.
This will be the final backport for the final 23.2 RC.

Release note: none.
Epic: REL-437

celiala committed:
clusterverion: define start of 24.1 dev versions and placeholder key

Release note: none.
Epic: REL-437.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111634)
<!-- Reviewable:end -->

celiala committed:
clusterversion: remove old guidance + check for >= 5 releases ago

Previously, we put this check in place to force the deletion of
old (no longer needed) gates. Since then:
- The team + number of version gates have scaled, so deleting gates
  as part of creating the next development gate is harder to do in
  a single atomic commit. Since 22.2, we've decoupled the deletion
  of old gates from the creation of the next development gate.
- Furthermore, with the introduction of VPrimordial gates, the
  check + error message is now incorrect: the error message says
  to delete 0.0-* gates; but these are VPrimordial gates, which
  should never be deleted.
- This rule assumed that versions older than 5 releases ago are no
  longer active. However, this rule is no longer always true, now
  that we've modified the scheduel for when we advance the minBinary
  (for VersionSkipping).

This commit removes this check, in order to prepare for the creation
of the 24.1 development version gate.

Release note: None.
Epic Link: REL-437.

celiala committed:
build: advance version to beta.1

This needs tp be merged onto release-23.2 right
before selecting the release candidate for 23.2.0-beta.1

Release note: None.
Epic: REL-437.

celiala committed:
clusterversion: set developmentBranch to false

This sets developmentBranch to false for the release branch.
We plan to backport this commit as soon as the branch is cut.

Release note: none.
Epic: REL-437.

celiala reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111634)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/clusterversion/keyed_versions.go` line 77 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111634#-NgpqF-c6dZLknn53Dw4:-NgpqF-c6dZLknn53Dw5:bcjvj66) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c69941f498f8e595769d3bc1f599ee6d35abd57/pkg/clusterversion/keyed_versions.go#L77)):*
> ```Go
> 		}
> 		// Add to existing entry.
> 		byRelease[n-1].vs = append(byRelease[n-1].vs, namedVersion)
> ```

Fixes: [#111889](https://github.com/cockroachdb/cockroach/issues/111889)

Removing this to remove [misleading failture](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/12045194?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildDeploymentsSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true) triggered while attempting to add V23_2 version gate (aka minting 23.2)
___
*[`pkg/cli/testdata/declarative-rules/deprules` line 3 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111634#-NgppmucDeffhZIoE0uF:-NgppmucDeffhZIoE0uG:b-pykuq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c69941f498f8e595769d3bc1f599ee6d35abd57/pkg/cli/testdata/declarative-rules/deprules#L3)):*
> ```
> dep
> ----
> debug declarative-print-rules 1000023.2 dep
> ```

ported over from https://github.com/cockroachdb/cockroach/pull/112347/files#diff-e1e782b33787152fcdb03ea0f667447cac80982e72e83241f2f0f8dc8a819a0e

h/t to @RaduBerinde for figuring how to adjust this test.
___
*[`pkg/sql/catalog/bootstrap/bootstrap_test.go` line 48 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/111634#-NgqJNx51STcTFeWX3n8:-NgqJNx51STcTFeWX3n9:b-eu0xz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/709afeb544a5ad9faa921fe1e068f396d01e82f5/pkg/sql/catalog/bootstrap/bootstrap_test.go#L48)):*
> ```Go
> 	var incumbent roachpb.Version
> 	for _, v := range clusterversion.ListBetween(earliest, latest) {
> 		if v.Major != incumbent.Major || v.Minor != incumbent.Minor {
> ```

Fixed buglet. Without this, was getting this error:

```
--- FAIL: TestSupportedReleases (0.01s)
    test_log_scope.go:170: test logs captured to: /tmp/cockroach/_tmp/403fccac927737279f86cb423ac2d596/logTestSupportedReleases2108764679
    test_log_scope.go:81: use -show-logs to present logs inline
    bootstrap_test.go:72:
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/sql/catalog/bootstrap/bootstrap_test.go:72
        	Error:      	Should be true
        	Test:       	TestSupportedReleases
        	Messages:   	expected supported releases map[1000023.1:{} 1000023.2-4:{}], actual map[1000023.1:{} 1000023.2:{} 1000023.2-4:{}]
        	            	see comments in test definition if this message appears
    panic.go:522: -- test log scope end --
FAIL
I231016 02:19:36.411234 1 (gostd) testmain.go:93  [-] 1  Test //pkg/sql/catalog/bootstrap:bootstrap_test exited with error code 1


```


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I think we should pull out the bug fixes in the tests  and backport those separately (even if they don't manifest on 23.2)

celiala commented:
Closing this PR.

@RaduBerinde and I synced offline:
Rather than temporarily toggling devBranch to false and then back to true on master, instead I will open a PR directly for release-23.2 to:
1) set devBranch to false
2) set version.txt to beta.1
This will use the timing guidelines in this PR, specificially: these two tasks should only be merged when/if we're ready to select a beta candidate.
